htlcswitch/iterator.go

```diff
@@ -194,32 +239,141 @@ func (p *OnionProcessor) DecodeHopIterator(r io.Reader, rHash []byte) (HopIterat
+// and failcodes upon subsequent invocations.
+//
+// NOTE: In order for the responses to be valid, the caller must guarantee that
+// the presented readers and rhashes *NEVER* deviate across invocations for the
+// same id.
+func (p *OnionProcessor) DecodeHopIterators(id []byte, rs []io.Reader,
+	rHashes [][]byte, incomingCltvs []uint32) ([]HopIterator,
+	[]lnwire.FailCode, error) {
 
-	onionPkt := &sphinx.OnionPacket{}
-	if err := onionPkt.Decode(r); err != nil {
+	batchSize := len(rs)
+
+	var (
+		onionPkts = make([]sphinx.OnionPacket, batchSize)
+		iterators = make([]HopIterator, batchSize)
+		failcodes = make([]lnwire.FailCode, batchSize)
+	)
+
+	tx := p.router.BeginTxn(id, batchSize)
+
+	for i, r := range rs {
+		onionPkt := &onionPkts[i]
+		err := onionPkt.Decode(r)
 		switch err {
+		case nil:
+			// success
+
 		case sphinx.ErrInvalidOnionVersion:
-			return nil, lnwire.CodeInvalidOnionVersion
+			failcodes[i] = lnwire.CodeInvalidOnionVersion
+			continue
+
 		case sphinx.ErrInvalidOnionKey:
-			return nil, lnwire.CodeInvalidOnionKey
+			failcodes[i] = lnwire.CodeInvalidOnionKey
+			continue
+
 		default:
 			log.Errorf("unable to decode onion packet: %v", err)
-			return nil, lnwire.CodeInvalidOnionKey
+			failcodes[i] = lnwire.CodeInvalidOnionKey
+			continue
+		}
+
+		err = tx.ProcessOnionPacket(
+			uint16(i), onionPkt, rHashes[i], incomingCltvs[i],
+		)
+		switch err {
+		case nil:
+			// success
+
+		case sphinx.ErrInvalidOnionVersion:
+			failcodes[i] = lnwire.CodeInvalidOnionVersion
+			continue
+
+		case sphinx.ErrInvalidOnionHMAC:
+			failcodes[i] = lnwire.CodeInvalidOnionHmac
+			continue
+
+		case sphinx.ErrInvalidOnionKey:
+			failcodes[i] = lnwire.CodeInvalidOnionKey
+			continue
+
+		default:
+			log.Errorf("unable to process onion packet: %v", err)
+			failcodes[i] = lnwire.CodeInvalidOnionKey
+			continue
 		}
 	}
 
+	// With that batch created, we will now attempt to write the shared
+	// secrets to disk. This operation will returns the set of indices that
+	// were detected as replays, and the computed sphinx packets for all
+	// indices that did not fail the above loop. Only indices that are not
+	// in the replay set should be considered valid, as they are
+	// opportunistically computed.
+	packets, replays, err := tx.Commit()
+	if err != nil {
+		log.Errorf("unable to process onion packet batch %x: %v",
+			id, err)
+		// If we failed to commit the batch to the secret share log, we
+		// will mark all not-yet-failed channels with a temporary
+		// channel failure and exit since we cannot proceed.
+		for i, fcode := range failcodes {
+			// Skip any indexes that already failed onion decoding.
+			if fcode != lnwire.CodeNone {
+				continue
+			}
+
+			log.Errorf("unable to process onion packet %x-%v",
+				id, i)
+			failcodes[i] = lnwire.CodeTemporaryChannelFailure
+		}
+
+		// TODO(conner): return real errors to caller so link can fail?
+		return iterators, failcodes, err
```

Comment:
Hmm, as is now, it seems that if _any_ of the HTLC's are actually a replay, then we won't proceed with forwarding the rest. Or is this error only returned in the case of a db level failure? (disk too full, or w/e)